### PR TITLE
Refactor binary

### PR DESCRIPTION
**Pull request by emricksinisonos**

### CODE REVIEW

File: core/src/ops/mod.rs [hunk 1]
```diff
@@ -11,6 +11,7 @@ pub mod macros;
 pub mod element_wise;
 #[macro_use]
 pub mod binary;
+//pub mod binary_new;
```
Comment (kali) [inline core/src/ops/mod.rs:14]: ...

File: data/src/tensor/indices.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+pub struct IndexIterator {
```
Comment (kali) [inline data/src/tensor/indices.rs:1]: is this better / different than ndarray::indexes() ?

File: linalg/src/arm64/arm64fp16/by_scalar.rs [hunk 3]
```diff
@@ -34,8 +34,183 @@ ew_impl_wrap!(
...
+        assert!(buf.len() % 16 == 0);
+        assert!(buf.len() > 0);
+        #[target_feature(enable = "fp16")]
+        unsafe fn run(buf: &mut[f16], s: f16) {
+            let len = buf.len();
+            let ptr = buf.as_ptr();
+            std::arch::asm!("
+            dup v0.8h, v0.h[0]
+            2:
+                ld1 {{v4.8h, v5.8h, v6.8h, v7.8h}}, [{ptr}]
```
Comment (kali) [inline linalg/src/arm64/arm64fp16/by_scalar.rs:53]: we need to give a shot at unrolling to amortize the loading latency, and prefetch too (but this is another task/PR)

File: linalg/src/frame/by_scalar.rs [hunk 4]
```diff
@@ -1,3 +1,61 @@
...
+
+use crate::{LADatum, element_wise::ElementWiseKer, LinalgFn};
+
+use super::{ElementWise, element_wise_helper::map_slice_with_alignment};
+
+
+/// Generic implementation struct that unify all by scalar kernels.
+/// A by scalar operation is an ElementWise operation with a scalar paramerer.
+#[derive(Debug, Clone, new)]
+pub struct ByScalarImpl<K, T>
```
Comment (kali) [inline linalg/src/frame/by_scalar.rs:13]: I'm slowing moving away from this pattern in the matmul code. We may need to discuss it. (mmm kernels are now represented by the instance of a struct, lazy-initialized const)

File: linalg/src/lib.rs [hunk 5]
```diff
@@ -183,13 +181,73 @@ lazy_static::lazy_static! {
...
+    };
+}
+
+pub fn bin_by_scalar(dt: DatumType, bin: BinOp) -> Option<Box<LinalgFn>> {
+    let map = BIN_BY_SCALAR_OPS.lock().unwrap();
+    map.get(&(bin, dt)).map(|it| (it)())
+}
+
+pub fn bin_unicast(dt: DatumType, bin: BinOp) -> Option<Box<LinalgFn>> {
+    let map = BIN_UNICAST_OPS.lock().unwrap();
```
Comment (kali) [inline linalg/src/lib.rs:238]: we should not need the mutex here, as we are initialized in a lazy_static. is this visible in profiles ? i think we may want to consider non-hashmap options (we can make BinOp repr(C) and use it as integer for instance, maybe something similar for dt ?)

**Review by kali — COMMENTED** [review]
Lets's hace a chat later today :)

File: data/src/tensor/indices.rs (hunk 2 above)
Comment (emricksinisonos) [reply data/src/tensor/indices.rs]: After re-checking, `ndarray::indexes()` does the job !

**Pull request merged**